src/apps/bridge/reportBuilder.cpp

```diff
@@ -0,0 +1,497 @@
+ * @param sensor_data Pointer to the sensor data for the report.
+ * @param sensor_data_size The size of the sensor data.
+ * @param samples_per_report The number of samples per report.
+ * @param sample_counter The current sample counter.
+ * @return A pointer to the newly created report builder element.
+ *
+ */
+report_builder_element_t* ReportBuilderLinkedList::newElement(uint64_t node_id, uint8_t sensor_type, void *sensor_data, uint32_t sensor_data_size, uint32_t samples_per_report, uint32_t sample_counter) {
+  report_builder_element_t *element = static_cast<report_builder_element_t *>(pvPortMalloc(sizeof(report_builder_element_t)));
+  configASSERT(element != NULL);
+  element->node_id = node_id;
+  element->sensor_type = sensor_type;
+  element->sensor_data = static_cast<void *>(pvPortMalloc(samples_per_report * sensor_data_size));
+  configASSERT(element->sensor_data != NULL);
+  memset(element->sensor_data, 0, samples_per_report * sensor_data_size);
+
+  addSensorDataToElement(element, sensor_type, sensor_data, sample_counter);
+
+  element->next = NULL;
+  element->prev = NULL;
+  return element;
+}
+
+void ReportBuilderLinkedList::freeElement(report_builder_element_t **element_pointer) {
+  if (element_pointer != NULL && *element_pointer != NULL) {
+    vPortFree((*element_pointer)->sensor_data); // Free the data first
+    vPortFree(*element_pointer);
+    *element_pointer = NULL;
+  }
+}
+
+bool ReportBuilderLinkedList::removeElement(report_builder_element_t *element) {
+  bool rval = false;
+  if (element != NULL) {
+    if (element->prev != NULL) {
+      element->prev->next = element->next;
+    } else {
+      head = element->next;
+    }
+    if (element->next != NULL) {
+      element->next->prev = element->prev;
+    }
+    freeElement(&element);
+    size--;
+    rval = true;
+  }
+  return rval;
+}
+
+/**
+ * @brief Adds sensor data to a specific element in the report builder linked list.
+ *
+ * This function takes a pointer to an element in the linked list and adds the sensor data to that element. The sensor data is added at the position specified by the sample counter.
+ *
+ * @param element Pointer to the element in the linked list.
+ * @param sensor_type The type of the sensor for the report.
+ * @param sensor_data Pointer to the sensor data for the report.
+ * @param sample_counter The current sample counter, indicating the position at which to add the sensor data.
+ */
+void ReportBuilderLinkedList::addSensorDataToElement(report_builder_element_t *element, uint8_t sensor_type, void *sensor_data, uint32_t sample_counter) {
+    switch (sensor_type) {
+      case AANDERAA_SENSOR_TYPE: {
+        // Back fill the sensor_data with NANs if we are not on the right sample count
+        uint32_t i;
+        for (i = 0; i < sample_counter; i++) {
+          memcpy(&(static_cast<aanderaa_aggregations_t *>(element->sensor_data))[i], &NAN_AGG, sizeof(aanderaa_aggregations_t));
```

Comment:
Nit: Not super necessary now, but maybe makes sense to have size_t sensor_data_size as an argument. Maybe just a  TODO comment is fine for now so you don't have to change all the calls.

Reply:
Talked IRL. Originally the `sensor_data_size` was an input, but then I realized that this function needs to know how to cast the void * sensor data and thus needs to already know the varying sensor data structs, so then it can extract the size of the structs from those instead of needing an input.